astropy/modeling/core.py

```diff
@@ -2236,774 +2240,1142 @@ def _join_operator(f, g):
+        exist. This method must be called to make the model usable by
+        fitting engines.
+        If oldnames=True, then parameters are named as in the original
+        implementation of compound models.
+        """
+        if self._parameters is not None:
+            # do nothing
+            return
+        if self._leaflist is None:
+            self._make_leaflist()
+        self._parameters_ = OrderedDict()
+        param_map = {}
+        self._param_names = []
+        for lindex, leaf in enumerate(self._leaflist):
+            for param_name in leaf.param_names:
+                param = getattr(leaf, param_name)
+                new_param_name = "{}_{}".format(param_name, lindex)
+                self.__dict__[new_param_name] = param
+                self._parameters_[new_param_name] = param
+                self._param_names.append(new_param_name)
+                param_map[new_param_name] = (lindex, param_name)
+        self._param_metrics = {}
+        self._param_map =  param_map
+        self._param_map_inverse = dict((v, k) for k, v in param_map.items())
+        self._initialize_slices()
+        self._initialize_constraints()
+
+
+    def _initialize_slices(self):
+        # TODO eliminate redundant code with core here and next two methods
+        param_metrics = self._param_metrics
+        total_size = 0
+
+        for name in self.param_names:
+            param = getattr(self, name)
+            value = param.value
+            param_size = np.size(value)
+            param_shape = np.shape(value)
+            param_slice = slice(total_size, total_size + param_size)
+            param_metrics[name] = {}
+            param_metrics[name]['slice'] = param_slice
+            param_metrics[name]['shape'] = param_shape
+            param_metrics[name]['size'] = param_size
+            total_size += param_size
+        self._parameters = np.empty(total_size, dtype=np.float64)
+
+    @staticmethod
+    def _recursive_lookup(branch, adict, key):
+        if isinstance(branch, CompoundModel):
+            return adict[key]
+        else:
+            return branch, key
+
+    def inputs_map(self):
+        """
+        Map the names of the inputs to this ExpressionTree to the inputs to the leaf models.
+        """
+        inputs_map = {}
+        if not isinstance(self.op, str):  # If we don't have an operator the mapping is trivial
+            return {inp: (self, inp) for inp in self.inputs}
+
+        elif self.op == '|':
+            if isinstance(self.left, CompoundModel):
+                l_inputs_map = self.left.inputs_map()
+            for inp in self.inputs:
+                if isinstance(self.left, CompoundModel):
+                    inputs_map[inp] = l_inputs_map[inp]
+                else:
+                    inputs_map[inp] = self.left, inp
+        elif self.op == '&':
+            if isinstance(self.left, CompoundModel):
+                l_inputs_map = left.inputs_map()
```

Comment:
`left` is an undefined name here, and there's two more undefined names in this file. Add test and fix code?
```
$ pyflakes astropy/modeling/core.py 
astropy/modeling/core.py:17: 'copyreg' imported but unused
astropy/modeling/core.py:34: 'astropy.utils.OrderedDescriptorContainer' imported but unused
astropy/modeling/core.py:40: '.utils.ExpressionTree' imported but unused
astropy/modeling/core.py:50: redefinition of unused 'indent' from line 30
astropy/modeling/core.py:51: redefinition of unused 'combine_labels' from line 40
astropy/modeling/core.py:51: redefinition of unused '_BoundingBox' from line 40
astropy/modeling/core.py:78: redefinition of unused '_model_oper' from line 58
astropy/modeling/core.py:1483: local variable 'name' is assigned to but never used
astropy/modeling/core.py:1777: local variable 'supplied_parvalues' is assigned to but never used
astropy/modeling/core.py:1923: local variable 'unit' is assigned to but never used
astropy/modeling/core.py:1973: local variable 'param_names' is assigned to but never used
astropy/modeling/core.py:2410: local variable 'linv' is assigned to but never used
astropy/modeling/core.py:2411: local variable 'rinv' is assigned to but never used
astropy/modeling/core.py:2896: undefined name 'left'
astropy/modeling/core.py:2898: undefined name 'right'
astropy/modeling/core.py:2988: local variable 'l_inputs_map' is assigned to but never used
astropy/modeling/core.py:2993: undefined name 'l_outputs_map'
astropy/modeling/core.py:3870: local variable 'inputs_are_quantity' is assigned to but never used
```

Reply:
Agree, this should be fixed

Reply:
fixed